File: InfiniteScrollViewDemo/CustomInfiniteIndicator.m
//
//  CustomInfiniteIndicator.m
//  InfiniteScrollViewDemo
//
//  Created by pronebird on 27/11/14.
//  Copyright (c) 2014 codeispoetry.ru. All rights reserved.
//

#import "CustomInfiniteIndicator.h"

static NSString *const kRotationAnimationKey = @"rotation";

@interface CustomInfiniteIndicator()

@property CAShapeLayer *outerCircle;
@property CAShapeLayer *innerCircle;
@property (readwrite) BOOL animating;

@end

@implementation CustomInfiniteIndicator

- (instancetype)initWithCoder:(NSCoder *)coder {
    if(self = [super initWithCoder:coder]) {
        [self _commonInit];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    if(self = [super initWithFrame:frame]) {
        [self _commonInit];
    }
    return self;
}

- (void)dealloc {
    [self _unregisterFromAppStateNotifications];
}

- (void)setThickness:(CGFloat)thickness {
    if(_thickness != thickness) {
        _thickness = thickness;
        
        self.innerCircle.lineWidth = thickness;
        self.outerCircle.lineWidth = thickness;
    }
}
- (void)setInnerColor:(UIColor *)innerColor {
    if(![_innerColor isEqual:innerColor]) {
        _innerColor = innerColor;
        
        self.innerCircle.strokeColor = innerColor.CGColor;
    }
}

- (void)setOuterColor:(UIColor *)outerColor {
    if(![_outerColor isEqual:outerColor]) {
        _outerColor = outerColor;
        
        self.outerCircle.strokeColor = outerColor.CGColor;
    }
}

- (void)startAnimating {
    if(self.animating) {
        return;
    }
    
    self.animating = YES;
    self.hidden = NO;
    [self _addAnimation];
}

- (void)stopAnimating {
    if(!self.animating) {
        return;
    }
    
    [self.layer removeAnimationForKey:kRotationAnimationKey];
    self.hidden = YES;
    self.animating = NO;
}

- (void)layoutSublayersOfLayer:(CALayer *)layer {
    [self _setupBezierPaths];
}

- (void)prepareForInterfaceBuilder {
    self.hidden = NO;
}

- (void)didMoveToWindow {
    [super didMoveToWindow];
    
    // CoreAnimation animations are removed when view goes offscreen.
    // So we have to restart them when view reappears.
    if(self.window) {
        [self _restartAnimationIfNeeded];
    }
}

#pragma mark - Private

- (void)_addAnimation {
    [self.layer addAnimation:[self _animation] forKey:kRotationAnimationKey];
}

- (void)_setupBezierPaths {
    CGPoint center = CGPointMake(self.bounds.size.width * 0.5, self.bounds.size.height * 0.5);
    CGFloat radius = self.bounds.size.width * 0.5 - self.thickness;
    UIBezierPath *ringPath = [UIBezierPath bezierPathWithArcCenter:center radius:radius startAngle:0 endAngle:M_PI * 2 clockwise:YES];
    UIBezierPath *quarterRingPath = [UIBezierPath bezierPathWithArcCenter:center radius:radius startAngle:-M_PI_4 endAngle:M_PI_2 - M_PI_4 clockwise:YES];
    
    self.outerCircle.path = ringPath.CGPath;
    self.innerCircle.path = quarterRingPath.CGPath;
}

- (void)_commonInit {
    [self _registerForAppStateNotifications];
    
    self.hidden = YES;
    self.backgroundColor = [UIColor clearColor];
    
    if(self.thickness < 1) {
#if TARGET_OS_TV
        self.thickness = 6;
#else
        self.thickness = 2;
#endif
    }
    
    if(!self.innerColor) {
        self.innerColor = self.tintColor;
    }
    
    if(!self.outerColor) {
        self.outerColor = [[UIColor grayColor] colorWithAlphaComponent:0.2];
    }
    
    CAShapeLayer *outerCircle = [CAShapeLayer layer];
    CAShapeLayer *innerCircle = [CAShapeLayer layer];
    
    outerCircle.strokeColor = self.outerColor.CGColor;
    outerCircle.fillColor = [UIColor clearColor].CGColor;
    outerCircle.lineWidth = self.thickness;
    
    innerCircle.strokeColor = self.innerColor.CGColor;
    innerCircle.fillColor = [UIColor clearColor].CGColor;
    innerCircle.lineWidth = self.thickness;
    
    self.innerCircle = innerCircle;
    self.outerCircle = outerCircle;
    
    [self _setupBezierPaths];
    
    [self.layer addSublayer:outerCircle];
    [self.layer addSublayer:innerCircle];
}

- (CABasicAnimation *)_animation {
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform.rotation"];
    animation.toValue = @(M_PI * 2);
    animation.duration = 1;
    animation.repeatCount = INFINITY;
    animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear];
    
    return animation;
}

- (void)_restartAnimationIfNeeded {
    if(self.animating && ![[self.layer animationKeys] containsObject:kRotationAnimationKey]) {
        [self _addAnimation];
    }
}

#pragma mark - Notifications

- (void)_registerForAppStateNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_restartAnimationIfNeeded) name:UIApplicationWillEnterForegroundNotification object:nil];
}

- (void)_unregisterFromAppStateNotifications {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: InfiniteScrollViewDemoSwift/CustomInfiniteIndicator.swift
//
//  CustomInfiniteIndicator.swift
//  InfiniteScrollViewDemoSwift
//
//  Created by pronebird on 5/3/15.
//  Copyright (c) 2015 pronebird. All rights reserved.
//

import UIKit

private let rotationAnimationKey = "rotation"

class CustomInfiniteIndicator: UIView {
    
    var thickness: CGFloat = 2
    var outerColor: UIColor? {
        didSet {
            updateColors()
        }
    }
    var innerColor: UIColor? {
        didSet {
            updateColors()
        }
    }

    private var animating = false
    private let innerCircle = CAShapeLayer()
    private let outerCircle = CAShapeLayer()
    private var startTime = CFTimeInterval(0)
    private var stopTime = CFTimeInterval(0)

    // MARK: - Public
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    deinit {
        unregisterFromAppStateNotifications()
    }
    
    override func layoutSublayers(of layer: CALayer) {
        super.layoutSublayers(of: layer)
        setupBezierPaths()
    }
    
    override func didMoveToWindow() {
        super.didMoveToWindow()
        
        if let _ = window {
            restartAnimationIfNeeded()
        }
    }

    override func tintColorDidChange() {
        if innerColor == nil {
            updateColors()
        }
    }

    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        if #available(iOS 13.0, tvOS 13.0, *) {
            if traitCollection.hasDifferentColorAppearance(comparedTo: previousTraitCollection) {
                updateColors()
            }
        }
    }

    func isAnimating() -> Bool {
        return animating
    }

    @objc func startAnimating() {
        guard !animating else { return }
        animating = true
        isHidden = false
        addAnimation()
    }

    @objc func stopAnimating() {
        guard animating else { return }
        animating = false
        isHidden = true
        removeAnimation()
    }
    
    // MARK: - Private
    
    private func commonInit() {
        registerForAppStateNotifications()
        
        isHidden = true
        backgroundColor = UIColor.clear

        outerCircle.fillColor = UIColor.clear.cgColor
        outerCircle.lineWidth = thickness

        innerCircle.fillColor = UIColor.clear.cgColor
        innerCircle.lineWidth = thickness

        updateColors()
        
        layer.addSublayer(outerCircle)
        layer.addSublayer(innerCircle)
    }

    private func updateColors() {
        let outerColor = self.outerColor ?? self.defaultOuterColor()
        let innerColor = self.innerColor ?? self.tintColor

        outerCircle.strokeColor = outerColor.cgColor
        innerCircle.strokeColor = innerColor?.cgColor
    }
    
    private func addAnimation() {
        let anim = animation()
        anim.timeOffset = stopTime - startTime
        
        layer.add(anim, forKey: rotationAnimationKey)
        
        startTime = layer.convertTime(CACurrentMediaTime(), from: nil)
    }
    
    private func removeAnimation() {
        layer.removeAnimation(forKey: rotationAnimationKey)
        
        stopTime = layer.convertTime(CACurrentMediaTime(), from: nil)
    }
    
    @objc func restartAnimationIfNeeded() {
        let anim = layer.animation(forKey: rotationAnimationKey)
        
        if animating && anim == nil {
            removeAnimation()
            addAnimation()
        }
    }
    
    private func registerForAppStateNotifications() {
        NotificationCenter.default.addObserver(self, selector: #selector(CustomInfiniteIndicator.restartAnimationIfNeeded), name: UIApplication.willEnterForegroundNotification, object: nil)
    }
    
    private func unregisterFromAppStateNotifications() {
        NotificationCenter.default.removeObserver(self)
    }
    
    private func animation() -> CABasicAnimation {
        let animation = CABasicAnimation(keyPath: "transform.rotation")
        animation.toValue = NSNumber(value: Double.pi * 2)
        animation.duration = 1
        animation.repeatCount = Float.infinity
        animation.timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.linear)
        
        return animation
    }
    
    private func setupBezierPaths() {
        let center = CGPoint(x: bounds.size.width * 0.5, y: bounds.size.height * 0.5)
        let radius = bounds.size.width * 0.5 - thickness
        let ringPath = UIBezierPath(arcCenter: center, radius: radius, startAngle: CGFloat(0), endAngle: CGFloat.pi * 2, clockwise: true)
        let quarterRingPath = UIBezierPath(arcCenter: center, radius: radius, startAngle: -CGFloat.pi / 4, endAngle: CGFloat.pi / 2 - CGFloat.pi / 4, clockwise: true)
        
        outerCircle.path = ringPath.cgPath
        innerCircle.path = quarterRingPath.cgPath
    }

    private func defaultOuterColor() -> UIColor {
        let defaultLightColor = UIColor.gray.withAlphaComponent(0.2)

        if #available(iOS 13.0, tvOS 13, *) {
            return UIColor { (traitCollection) -> UIColor in
                switch traitCollection.userInterfaceStyle {
                case .light, .unspecified:
                    return defaultLightColor
                case .dark:
                    return UIColor.white.withAlphaComponent(0.5)
                @unknown default:
                    fatalError()
                }
            }
        } else {
            return defaultLightColor
        }
    }

}
